### PR TITLE
DOXYGEN: Adding description for request routines.

### PR DESCRIPTION
**Pull request by shamisp**

### CODE REVIEW

**Comment by shamisp** [comment]
autobot build

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1117/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/539/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1126/ for details.

**Comment by shamisp** [comment]
@yosefe please review

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -1269,8 +1272,17 @@ int ucp_request_is_completed(void *request);
  *
  * @param [in]  request      Non-blocking request to release.
  *
- * @note If the request is not completed yet, it will actually be released when
- *       completed.
+ * This routine marks and potentially releases back to the library the
+ * non-blocking request. If the request is already is already completed or
+ * canceled state it is released the resources associated with the request are
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: there is no "canceled" state, only "completed"

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -1269,8 +1272,17 @@ int ucp_request_is_completed(void *request);
...
+ * returned back to the library.  If the request in any other stated it is
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: last sentence needs a rephrase?

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -1269,8 +1272,17 @@ int ucp_request_is_completed(void *request);
...
+ * marked as a "ready to release" and it will be released once it enters
+ * completed and canceled states.
+ *
+ * @todo I think this release semantics is a bit confusing. I would suggest to
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: we need to do it to satisfy MPI semantics. you can say that we only release the handle here, and do not affect the course of the operation itself.

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -1282,11 +1294,22 @@ void ucp_request_release(void *request);
...
+ * with the communication request.  If the request is completed successfully,
+ * the @ref ucp_send_callback_t "send" or @ref ucp_tag_recv_callback_t
+ * "receive" completion callbacks (based on the type of the request) will be
+ * called with the @a status argument of the callback set to UCS_OK, and in a
+ * case it is canceled the @a status argument is set to UCS_ERR_CANCELED.  It is
+ * important to note that in order to release the request back to the library
+ * the application is responsible to call @ref ucp_request_release
+ * "ucp_request_release()".
+ *
+ * @note This routine does not undo any memory alternations associated with the
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: a bit confusing - a memory operation can either be completed or canceled. Why would it be meaningful to "undo" it?

File: src/ucp/api/ucp.h [hunk 5]
```diff
@@ -1282,11 +1294,22 @@ void ucp_request_release(void *request);
...
+ * communication operations.  This is a local operation that resets the status if
+ * the request and removes it from internal queues (e.g matching, pending,
+ * posted queues, etc.)
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: we will also have protocol to cancel rendezvous, which is not really a "local" operation (although according to MPI spec it still is)

File: src/ucp/api/ucp.h (hunk 4 above)
Comment (shamisp) [inline src/ucp/api/ucp.h]: So we do have completed and canceled states ?

File: src/ucp/api/ucp.h (hunk 5 above)
Comment (shamisp) [inline src/ucp/api/ucp.h]: We don't really send AM to the remote end, right ?

**Comment by shamisp** [comment]
@yosefe - we have to talk about the defined states. Most of the confusion is associated with the fact that we don't have well defined state for the request.

File: src/ucp/api/ucp.h (hunk 4 above)
Comment (yosefe) [inline src/ucp/api/ucp.h]: rma cannot be canceled

File: src/ucp/api/ucp.h (hunk 5 above)
Comment (yosefe) [inline src/ucp/api/ucp.h]: actually, yes.

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1135/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/552/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h (hunk 5 above)
Comment (shamisp) [inline src/ucp/api/ucp.h]: I see. I would have to introduce more generic description.

File: src/ucp/api/ucp.h (hunk 4 above)
Comment (shamisp) [inline src/ucp/api/ucp.h]: depends how it was implemented ? If it goes through some protocol, potentially it can be canceled.

**Comment by shamisp** [comment]
:+1:

File: src/ucp/api/ucp.h (hunk 4 above)
Comment (shamisp) [inline src/ucp/api/ucp.h]: I will remote the note for now

File: src/ucp/api/ucp.h (hunk 3 above)
Comment (shamisp) [inline src/ucp/api/ucp.h]: I will keep this todo. It does not appear in the generated documents

File: src/ucp/api/ucp.h (hunk 2 above)
Comment (shamisp) [inline src/ucp/api/ucp.h]: will be fixed

**Comment by shamisp** [comment]
@yosefe , please take a look.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1141/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/558/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/559/ for details (Mellanox internal link).

**Pull request merged**